library/core/src/ptr/mod.rs

```diff
@@ -427,6 +427,237 @@ pub use unique::Unique;
 mod const_ptr;
 mod mut_ptr;
 
+/// The list of address spaces.
+///
+/// See [`AddrspacePtr`] for using pointers with an explicit associated address space.
+///
+/// This module lists the address spaces for the current target.
+#[unstable(feature = "ptr_addrspace", issue = "none")]
+pub mod addrspace {
+    /// The default address space.
+    ///
+    /// On most targets, a raw pointer (`*mut/const T`) implicitly points into the generic address space.
+    #[unstable(feature = "ptr_addrspace", issue = "none")]
+    pub const GENERIC: u32 = 0;
+
+    /// The address space corresponding to VRAM on GPUs.
+    ///
+    /// It is readable and writable from the GPU and CPU.
+    /// Pointers into this address space can be casted to and from the generic address space.
+    #[unstable(feature = "ptr_addrspace", issue = "none")]
+    #[cfg(any(doc, target_arch = "amdgpu", target_arch = "nvptx64"))]
+    #[doc(cfg(any(target_arch = "amdgpu", target_arch = "nvptx64")))]
+    pub const GLOBAL: u32 = 1;
+
+    /// The address space shared between threads within the same workgroup on GPUs.
+    ///
+    /// It is readable and writable from the GPU.
+    /// Pointers into this address space can be casted to and from the generic address space.
+    #[unstable(feature = "ptr_addrspace", issue = "none")]
+    #[cfg(any(doc, target_arch = "amdgpu", target_arch = "nvptx64"))]
+    #[doc(cfg(any(target_arch = "amdgpu", target_arch = "nvptx64")))]
+    pub const WORKGROUP: u32 = 3;
+
+    /// The address space for constant memory on GPUs.
+    ///
+    /// It is readable from the GPU but not writable.
+    /// Pointers into this address space can be casted to and from the generic address space.
+    #[unstable(feature = "ptr_addrspace", issue = "none")]
+    #[cfg(any(doc, target_arch = "amdgpu", target_arch = "nvptx64"))]
+    #[doc(cfg(any(target_arch = "amdgpu", target_arch = "nvptx64")))]
+    pub const CONST: u32 = 4;
```

Comment:
What are these random constants, where do they come from? If they are LLVM-specific, then they should not show up here -- after all, what should the cranelift/GCC backend or Miri do with them?

Reply:
They're essentially arbitrary identifiers for identifying what address space a nonlocal pointer points into. 

The most correct representation would probably be a `#[non_exhaustive]` enum, but I don't know how difficult that would be to introspect in the compiler compared to a simple const generic.

Reply:
Arbitrary identifiers defined by who? Who is the receiver that interprets these numbers? Are they fixed by the target ABI or so, or is it just something internal to the LLVM codegen backends for these targets?

Reply:
The names are defined by the OpenCL standard, "with some additions." The numbers "correspond to target architecture specific LLVM address space numbers used in LLVM IR" and are defined by the AMDGPU target.

https://llvm.org/docs/AMDGPUUsage.html#address-spaces